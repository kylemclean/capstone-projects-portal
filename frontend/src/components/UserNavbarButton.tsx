import * as React from "react"
import { Logout } from "@mui/icons-material"
import {
    Box,
    Tooltip,
    ListItemIcon,
    IconButton,
    Menu,
    MenuItem,
    Divider,
    Link,
} from "@mui/material"
import { useContext, useState, MouseEventHandler } from "react"
import { Link as RouterLink } from "react-router-dom"
import GlobalContext from "../global-state/context"
import UserAvatar from "./UserAvatar"
import { portalApiInstance } from "../api/portal-api"

export default function UserNavbarButton(): JSX.Element | null {
    const { globalState, dispatch } = useContext(GlobalContext)

    const [anchorEl, setAnchorEl] = useState<HTMLElement | null>(null)
    const open = Boolean(anchorEl)
    const handleClick: MouseEventHandler<HTMLButtonElement> = (event) => {
        setAnchorEl(event.currentTarget)
    }
    const handleClose = () => {
        setAnchorEl(null)
    }

    if (!globalState.currentUser.logged_in) return null

    return (
        <>
            <Box
                sx={{
                    display: "flex",
                    alignItems: "center",
                    textAlign: "center",
                }}
            >
                <Tooltip
                    title={`Logged in as ${globalState.currentUser?.name}`}
                >
                    <IconButton
                        onClick={handleClick}
                        size="small"
                        sx={{ ml: 2 }}
                    >
                        <UserAvatar user={globalState.currentUser} />
                    </IconButton>
                </Tooltip>
            </Box>
            <Menu
                anchorEl={anchorEl}
                open={open}
                onClose={handleClose}
                onClick={handleClose}
                PaperProps={{
                    elevation: 0,
                    sx: {
                        overflow: "visible",
                        filter: "drop-shadow(0px 2px 8px rgba(0,0,0,0.32))",
                        mt: 1.5,
                        "& .MuiAvatar-root": {
                            width: 32,
                            height: 32,
                            ml: -0.5,
                            mr: 1,
                        },
                        "&:before": {
                            content: '""',
                            display: "block",
                            position: "absolute",
                            top: 0,
                            right: 14,
                            width: 10,
                            height: 10,
                            bgcolor: "background.paper",
                            transform: "translateY(-50%) rotate(45deg)",
                            zIndex: 0,
                        },
                    },
                }}
                transformOrigin={{ horizontal: "right", vertical: "top" }}
                anchorOrigin={{ horizontal: "right", vertical: "bottom" }}
            >
                {/* General menu items */}
                <MenuItem
                    component={RouterLink}
                    to={`/profiles/${globalState.currentUser.id}`}
                >
                    Profile
                </MenuItem>
                <Divider />

                {/* Admin menu items */}
                {globalState.currentUser.is_superuser && [
                    <MenuItem
                        key="admin"
                        component={Link}
                        href={process.env.REACT_APP_ADMIN_URL ?? "/admin/"}
                    >
                        Admin
                    </MenuItem>,
                    <MenuItem key="import" component={RouterLink} to="/import">
                        Import data
                    </MenuItem>,
                    <Divider key="divider" />,
                ]}

                {/* Account menu items */}
                <MenuItem component={RouterLink} to="/settings">
                    Settings
                </MenuItem>
                <MenuItem onClick={() => portalApiInstance.logout(dispatch)}>
                    <ListItemIcon>
                        <Logout fontSize="small" />
                    </ListItemIcon>
                    Log out
                </MenuItem>
            </Menu>
        </>
    )
}
