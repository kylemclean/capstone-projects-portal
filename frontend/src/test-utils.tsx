import * as React from "react"
import { useLocation, useParams } from "react-router-dom"
import { screen, waitFor } from "@testing-library/react"
import { useCurrentUser } from "./hooks"

/**
 * Component to display the login status of the user
 */
export const LoginStatus = (): JSX.Element => {
    const currentUser = useCurrentUser()
    return (
        <div>
            {currentUser.logged_in
                ? `Logged in as ${currentUser.name}`
                : "Not logged in"}
        </div>
    )
}

/**
 * Mock home page that shows the login status
 */
export const MockHomePage = (): JSX.Element => (
    <div>
        Home page
        <LoginStatus />
    </div>
)

/**
 * Mock profile page that shows user ID
 */
export const MockProfilePage = (): JSX.Element => {
    const { id } = useParams<{ id: string }>()
    return <div>Profile page for {id}</div>
}

/**
 * Component that displays each query parameter in a list of "key=value" pairs
 * Requires the use of the `useLocation` hook
 */
export const QueryParamsDisplay = (): JSX.Element => {
    const { search } = useLocation()
    const params = new URLSearchParams(search)
    return (
        <div>
            <ul>
                {Array.from(params.entries()).map(([key, value]) => (
                    <li key={key}>
                        {key}={value}
                    </li>
                ))}
            </ul>
        </div>
    )
}

/**
 * Waits for a query parameter `name` to be set to `value`, or unset.
 * Requires a `QueryParamsDisplay` component to be rendered on the screen.
 * @param name the name of the query parameter
 * @param value the value of the query parameter, or `null` if it should not be present
 */
export const waitForQueryParam = async (
    name: string,
    value: string | null
): Promise<void> =>
    waitFor(() => {
        const text = `${name}=${value}`
        if (value !== null) screen.getByText(text)
        else expect(screen.queryByText(text)).toBeNull()
    })

/**
 * Mock login page that shows where it will redirect to after login
 */
export const MockLoginPage = (): JSX.Element => {
    const location = useLocation()
    const next = new URLSearchParams(location.search).get("next") ?? "/"
    return <div>Login page, will redirect to {next}</div>
}
