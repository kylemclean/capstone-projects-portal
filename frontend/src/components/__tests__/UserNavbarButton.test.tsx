import { fireEvent, render, screen } from "@testing-library/react"
import "@testing-library/jest-dom"
import { MemoryRouter, Route } from "react-router-dom"
import GlobalStateProvider from "../../global-state/provider"
import type State from "../../global-state/state"
import type CurrentUserInfo from "../../models/current-user-info"
import { LoginStatus, MockProfilePage } from "../../test-utils"
import UserNavbarButton from "../UserNavbarButton"

/**
 * Renders the UserNavbarButton with a GlobalStateProvider and router.
 * @param initialState initial state to set for the GlobalStateProvider. If undefined, the state will be the default.
 * @returns the return value of the render function
 */
const renderUserNavbarButton = (
    initialState: State & { currentUser: { logged_in: true } }
) =>
    render(
        <GlobalStateProvider initialState={initialState}>
            <MemoryRouter initialEntries={["/"]}>
                <Route path="/" exact>
                    <UserNavbarButton />
                    <LoginStatus />
                </Route>
                <Route path="/profiles/:id" exact component={MockProfilePage} />
                <Route path="/settings" exact>
                    Settings page
                </Route>
                <Route path="/import" exact>
                    Import page
                </Route>
            </MemoryRouter>
        </GlobalStateProvider>
    )

const queryMenuButtons = () => {
    const profileButton = screen.queryByRole("menuitem", { name: "Profile" })
    const adminButton = screen.queryByRole("menuitem", { name: "Admin" })
    const importButton = screen.queryByRole("menuitem", { name: "Import data" })
    const settingsButton = screen.queryByRole("menuitem", { name: "Settings" })
    const logOutButton = screen.queryByRole("menuitem", { name: "Log out" })
    return {
        profileButton,
        adminButton,
        importButton,
        settingsButton,
        logOutButton,
    }
}

const regularUser: CurrentUserInfo = {
    logged_in: true,
    is_superuser: false,
    has_password: true,
    id: "1fc66da3-5f63-4dd8-aaac-425a18804cdc",
    name: "Guy User",
    image: "https://example.com/image.jpg",
}

const superUser: CurrentUserInfo = {
    logged_in: true,
    is_superuser: true,
    has_password: true,
    id: "1fc66da3-5f63-4dd8-aaac-425a18804cde",
    name: "Super User",
    image: "https://example.com/image.jpg",
}

it("matches snapshot for regular user", async () => {
    const { container } = renderUserNavbarButton({ currentUser: regularUser })
    expect(container).toMatchSnapshot()
})

it("matches snapshot for superuser", async () => {
    const { container } = renderUserNavbarButton({ currentUser: superUser })
    expect(container).toMatchSnapshot()
})

it("matches snapshot for user without image", async () => {
    const { container } = renderUserNavbarButton({
        currentUser: {
            logged_in: true,
            is_superuser: false,
            has_password: true,
            id: "1fc66da3-5f63-4dd8-aaac-425a18804cdf",
            name: "Jeff User",
            image: undefined,
        },
    })
    expect(container).toMatchSnapshot()
})

it("has working profile button in menu", async () => {
    renderUserNavbarButton({ currentUser: regularUser })

    // Click UserNavbarButton
    const button = screen.getByRole("button")
    fireEvent.click(button)

    // Assert that profile button is present
    const { profileButton } = queryMenuButtons()
    expect(profileButton).not.toBeNull()

    // Click profile button
    fireEvent.click(profileButton!)

    // Assert that profile page for current user is rendered
    screen.getByText(`Profile page for ${regularUser.id}`)
})

it("has working settings button in menu", async () => {
    renderUserNavbarButton({ currentUser: regularUser })

    // Click UserNavbarButton
    const button = screen.getByRole("button")
    fireEvent.click(button)

    // Assert that settings button is present
    const { settingsButton } = queryMenuButtons()
    expect(settingsButton).not.toBeNull()

    // Click settings button
    fireEvent.click(settingsButton!)

    // Assert that settings page is rendered
    screen.getByText("Settings page")
})

it("has working log out button in menu", async () => {
    localStorage.setItem("token", "OLDTOKEN")
    renderUserNavbarButton({ currentUser: regularUser })

    // Click UserNavbarButton
    const button = screen.getByRole("button")
    fireEvent.click(button)

    // Assert that log out button is present
    const { logOutButton } = queryMenuButtons()
    expect(logOutButton).not.toBeNull()

    // Click log out button
    fireEvent.click(logOutButton!)

    // Assert that the user is logged out
    screen.getByText("Not logged in")

    // Assert that the token is removed from local storage
    expect(localStorage.getItem("token")).toBeNull()
})

it("has working admin button for superusers", async () => {
    renderUserNavbarButton({ currentUser: superUser })

    // Click UserNavbarButton
    const button = screen.getByRole("button")
    fireEvent.click(button)

    // Assert that admin button is present
    const { adminButton } = queryMenuButtons()
    expect(adminButton).not.toBeNull()

    // Assert that admin button leads to the admin page
    expect(adminButton!).toHaveAttribute(
        "href",
        import.meta.env.VITE_ADMIN_URL ?? "/admin/"
    )
})

it("has working import data button for superusers", async () => {
    renderUserNavbarButton({ currentUser: superUser })

    // Click UserNavbarButton
    const button = screen.getByRole("button")
    fireEvent.click(button)

    // Assert that import data button is present
    const { importButton } = queryMenuButtons()
    expect(importButton).not.toBeNull()

    // Click import data button
    fireEvent.click(importButton!)

    // Assert that import data page is rendered
    screen.getByText("Import page")
})

it("does not have an import data button for regular users", async () => {
    renderUserNavbarButton({ currentUser: regularUser })

    // Click UserNavbarButton
    const button = screen.getByRole("button")
    fireEvent.click(button)

    // Assert that import data button is not present
    const { importButton } = queryMenuButtons()
    expect(importButton).toBeNull()
})

it("does not have an admin button for regular users", async () => {
    renderUserNavbarButton({ currentUser: regularUser })

    // Click UserNavbarButton
    const button = screen.getByRole("button")
    fireEvent.click(button)

    // Assert that admin button is not present
    const { adminButton } = queryMenuButtons()
    expect(adminButton).toBeNull()
})
