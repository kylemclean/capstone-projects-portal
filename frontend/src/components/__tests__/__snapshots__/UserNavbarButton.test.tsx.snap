// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot for regular user 1`] = `
<div>
  <div
    class="MuiBox-root css-5nwj3y"
  >
    <button
      aria-label="Logged in as Guy User"
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-137ul40-MuiButtonBase-root-MuiIconButton-root"
      data-mui-internal-clone-element="true"
      tabindex="0"
      type="button"
    >
      <div
        class="MuiAvatar-root MuiAvatar-circular css-1wlk0hk-MuiAvatar-root"
      >
        <img
          alt="Guy User"
          class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
          src="https://example.com/image.jpg"
        />
      </div>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
  <div>
    Logged in as Guy User
  </div>
</div>
`;

exports[`matches snapshot for superuser 1`] = `
<div>
  <div
    class="MuiBox-root css-5nwj3y"
  >
    <button
      aria-label="Logged in as Super User"
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-137ul40-MuiButtonBase-root-MuiIconButton-root"
      data-mui-internal-clone-element="true"
      tabindex="0"
      type="button"
    >
      <div
        class="MuiAvatar-root MuiAvatar-circular css-1wlk0hk-MuiAvatar-root"
      >
        <img
          alt="Super User"
          class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
          src="https://example.com/image.jpg"
        />
      </div>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
  <div>
    Logged in as Super User
  </div>
</div>
`;

exports[`matches snapshot for user without image 1`] = `
<div>
  <div
    class="MuiBox-root css-5nwj3y"
  >
    <button
      aria-label="Logged in as Jeff User"
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-137ul40-MuiButtonBase-root-MuiIconButton-root"
      data-mui-internal-clone-element="true"
      tabindex="0"
      type="button"
    >
      <div
        class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-2s90m6-MuiAvatar-root"
      >
        J
      </div>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
  <div>
    Logged in as Jeff User
  </div>
</div>
`;
